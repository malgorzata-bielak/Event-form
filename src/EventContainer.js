import React from "react";
import EventForm from "./EventForm";

export default class EventContainer extends React.Component {
  state = {
    firstname: "",
    lastname: "",
    email: "",
    date: new Date(),
    error: ""
  };

  onChange = e => {
    const { value, id } = e.target;
    this.setState({
      [id]: value.trim()
    });
  };

  onDateChange = date => {
    this.setState({
      date
    });
  };

  onSubmit = e => {
    const { firstname, lastname, email, date } = this.state;
    e.preventDefault();

    if (!firstname || !lastname || !email || !date) {
      this.setState(() => ({ error: "Please provide all the necessary information" }));
    } else if (email && !email.match(/^[a-z0-9]+[a-z0-9.-_]+@[a-z0-9.-]+[a-z]$/i)) {
      this.setState(() => ({ error: "Please enter a valid email address" }));
    } else {
      this.setState(() => ({ error: "" }));
    }
  };

  render() {
    return (
      <>
        {this.state.error ? <p>{`${this.state.error}`}</p> : ""}
        <EventForm
          date={this.state.date}
          onChange={this.onChange}
          onDateChange={this.onDateChange}
          onSubmit={this.onSubmit}
        />
      </>
    );
  }
}
