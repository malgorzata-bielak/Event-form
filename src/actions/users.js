export const saveUserData = user => ({
  type: "SAVE_USER_DATA",
  user
});
