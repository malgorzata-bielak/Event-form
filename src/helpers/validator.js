export const isNameValid = name => name && name.match(/^[a-z-\s\u00C0-\u024F\u1E00-\u1EFF]*$/i);

export const isEmailValid = email =>
  email && email.match(/^[a-z0-9]+[a-z0-9-._]*@[a-z0-9.-]+[a-z]$/i);
