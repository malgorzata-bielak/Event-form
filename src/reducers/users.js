const usersReducerDefaultState = [];

const usersReducer = (state = usersReducerDefaultState, action) => {
  switch (action.type) {
    case "SAVE_USER_DATA":
      return [...state, action.user];
    default:
      return state;
  }
};

export default usersReducer;
