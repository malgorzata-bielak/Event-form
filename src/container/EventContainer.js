import React from "react";
import { connect } from "react-redux";

import EventForm from "../components/EventForm";
import FeedbackMessage from "../components/FeedbackMessage";
import FormWrapper from "../components/FormWrapper";
import { saveUserData } from "../actions/users";
import { addUser } from "../services/service";
import { isNameValid, isEmailValid } from "../validators/validator";

export class EventContainer extends React.Component {
  state = {
    firstname: "",
    lastname: "",
    email: "",
    date: new Date(),
    validationError: "",
    databaseResponse: "",
    isSaving: false
  };

  onChange = e => {
    const { value, id } = e.target;
    this.setState({
      [id]: value.trim(),
      databaseResponse: ""
    });
  };

  onDateChange = date => {
    this.setState({
      date
    });
  };

  backendResponse = databaseResponse => {
    this.setState({
      databaseResponse,
      isSaving: false
    });
  };

  onSubmit = e => {
    e.preventDefault();
    const { firstname, lastname, email, date } = this.state;

    if (!firstname || !lastname || !email || !date) {
      return this.setState({
        validationError: "Please provide all the necessary information"
      });
    }

    if (!isNameValid(firstname)) {
      return this.setState({ validationError: "Please enter a valid first name" });
    }

    if (!isNameValid(lastname)) {
      return this.setState({ validationError: "Please enter a valid last name" });
    }

    if (!isEmailValid(email)) {
      return this.setState({ validationError: "Please enter a valid email address" });
    }

    this.setState({ validationError: "", isSaving: true, databaseResponse: "" });

    const user = {
      firstname,
      lastname,
      email: email.toLowerCase(),
      date: date.valueOf()
    };

    addUser({
      saveUserData,
      user,
      backendResponse: this.backendResponse
    });
  };

  render() {
    return (
      <FormWrapper>
        <FeedbackMessage
          validationError={this.state.validationError}
          isSaving={this.state.isSaving}
          databaseResponse={this.state.databaseResponse}
        />
        <EventForm
          date={this.state.date}
          onChange={this.onChange}
          onDateChange={this.onDateChange}
          onSubmit={this.onSubmit}
        />
      </FormWrapper>
    );
  }
}

export default connect(
  undefined,
  { saveUserData }
)(EventContainer);
