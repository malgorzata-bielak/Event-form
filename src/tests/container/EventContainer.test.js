import React from "react";
import { shallow } from "enzyme";
import MockDate from "mockdate";

import { EventContainer } from "../../container/EventContainer";
import { getChildById } from "../testData/helpers";
import { users, invalidData } from "../testData/users";

beforeAll(() => {
  MockDate.set("5/23/2019");
});

afterAll(() => {
  MockDate.reset();
});

test("should render EventContainer correctly", () => {
  const wrapper = shallow(<EventContainer />);
  expect(wrapper).toMatchSnapshot();
});

test("should set firstname on input change", () => {
  const id = "firstname";
  const value = "James";
  const wrapper = shallow(<EventContainer />);

  const form = wrapper.find("EventForm");
  const label = getChildById(form, id);
  const input = getChildById(label, id);

  input.simulate("change", {
    target: { value, id }
  });
  expect(wrapper.state("firstname")).toBe(value);
  expect(wrapper.state("databaseResponse")).toBe("");
});

test("should set lastname on input change", () => {
  const id = "lastname";
  const value = "Bond";
  const wrapper = shallow(<EventContainer />);

  const form = wrapper.find("EventForm");
  const label = getChildById(form, id);
  const input = getChildById(label, id);

  input.simulate("change", {
    target: { value, id }
  });
  expect(wrapper.state("lastname")).toBe(value);
  expect(wrapper.state("databaseResponse")).toBe("");
});

test("should set email on input change", () => {
  const id = "email";
  const value = "james.bond@gmail.com";
  const wrapper = shallow(<EventContainer />);

  const form = wrapper.find("EventForm");
  const label = getChildById(form, id);
  const input = getChildById(label, id);

  input.simulate("change", {
    target: { value, id }
  });
  expect(wrapper.state("email")).toBe(value);
  expect(wrapper.state("databaseResponse")).toBe("");
});

test("should set date on input change", () => {
  const id = "date-picker";
  const value = 1558523468533;
  const wrapper = shallow(<EventContainer />);

  const form = wrapper.find("EventForm");
  const label = getChildById(form, id);

  label.find("Styled(DatePicker)").simulate("change", value);
  expect(wrapper.state("date")).toBe(value);
});

test("should change state with expected informations if valid form submission", () => {
  const wrapper = shallow(<EventContainer />);
  wrapper.setState({ ...users[0] });

  const form = wrapper.find("EventForm");
  form.simulate("submit", { preventDefault: () => {} });

  expect(wrapper.state("validationError")).toBe("");
  expect(wrapper.state("databaseResponse")).toBe("");
  expect(wrapper.state("isSaving")).toBe(true);
});

test("should render error if any form field is empty", () => {
  const wrapper = shallow(<EventContainer />);
  wrapper.setState({ ...invalidData[0] });

  const form = wrapper.find("EventForm");
  form.simulate("submit", { preventDefault: () => {} });

  expect(wrapper.state("validationError")).toBe("Please provide all the necessary information");
});

test("should render error if firstname is invalid", () => {
  const wrapper = shallow(<EventContainer />);
  wrapper.setState({ ...invalidData[1] });

  const form = wrapper.find("EventForm");
  form.simulate("submit", { preventDefault: () => {} });

  expect(wrapper.state("validationError")).toBe("Please enter a valid first name");
});

test("should render error if lastname is invalid", () => {
  const wrapper = shallow(<EventContainer />);
  wrapper.setState({ ...invalidData[2] });

  const form = wrapper.find("EventForm");
  form.simulate("submit", { preventDefault: () => {} });

  expect(wrapper.state("validationError")).toBe("Please enter a valid last name");
});

test("should render error if email is invalid", () => {
  const wrapper = shallow(<EventContainer />);
  wrapper.setState({ ...invalidData[3] });

  const form = wrapper.find("EventForm");
  form.simulate("submit", { preventDefault: () => {} });

  expect(wrapper.state("validationError")).toBe("Please enter a valid email address");
});
