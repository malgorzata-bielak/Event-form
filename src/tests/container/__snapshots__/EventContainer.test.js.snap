// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EventContainer correctly 1`] = `
<Styled(div)>
  <FeedbackMessage
    databaseResponse=""
    isSaving={false}
    validationError=""
  />
  <EventForm
    date={2019-05-22T22:00:00.000Z}
    onChange={[Function]}
    onDateChange={[Function]}
    onSubmit={[Function]}
  />
</Styled(div)>
`;
